django/db/models/query.py

```diff
@@ -963,7 +963,25 @@ def update_or_create(self, defaults=None, **kwargs):
                 return obj, created
             for k, v in resolve_callables(defaults):
                 setattr(obj, k, v)
-            obj.save(using=self.db)
+
+            update_fields = set(defaults)
+            concrete_field_names = self.model._meta._non_pk_concrete_field_names
+
+            # `update_fields` does not support non-concrete fields (#31382)
+            if concrete_field_names.issuperset(update_fields):
+                # Add fields which are set on pre_save, e.g. `auto_now` fields.
+                # This is to maintain backward compatibility as these fields are not
+                # updated if not specified explicitly in the `update_fields` list.
+                for field in self.model._meta.local_concrete_fields:
+                    if not field.primary_key:
+                        if getattr(obj, field.attname) != field.pre_save(obj, False):
```

Comment:
Arf, this is also not optimal either.

`pre_save` can have side-effects, like `django.db.models.fields.files.FileField.pre_save` does 😕 
We probably don't want to trigger those here.

I mean, serendipitously it would work for the `FileField` because even if the returned value is still the same (so we don't add the `field.name` to `updated_fields`), we actually triggered the side-effect committing the file 😂 
However, that seems pretty brittle 😅 

I'm not sure what the cleanest/Djangoest approach would be here 🤔 
We could add an attribute on the Field class, like `Field.has_pre_save: bool`, but that creates a precedent and users/libs must update their code accordingly.
But at least, we would know _for sure_ which fields need to be added and which don't.

Any other suggestion is very welcome!

Reply:
I am now comparing the bytecode of the functions which allows us to see if it has a custom `pre_save` without having to call the `pre_save`'s themselves.

I have never used this before, it seems to work. Might be a little unusual However the tests are passing and it doesn't require any one to update their code. :shrug: 

The `FileField` will be included in the update_fields but I guess that is ok, as before we were calling the pre_save on all fields as they were using the standard save - so it is kind of existing behaviour.

Reply:
My recollection is a touch fuzzy/uncertain, but you _may_ be able to get away with _not_ looking at at the deep internals (`co_code`), and just comparing `Field.pre_save is field_instance.__class__.pre_save`
Though to caveat that, there may be circumstances (hello multiprocesses?) where that doesn't work?

Reply:
@kezabelle that's so much nicer thank you :star: